ambry-tools/src/main/java/com.github.ambry/tools/perf/rest/NettyPerfClient.java

```diff
@@ -424,10 +458,33 @@ protected void shutdown() {
    * Blocking function to wait on the NettyPerfClient shutting down.
    * @throws InterruptedException
    */
-  protected void awaitShutdown() throws InterruptedException {
+  private void awaitShutdown() throws InterruptedException {
     shutdownLatch.await();
   }
 
+  /**
+   * A periodic updater that adjusts sleep time for connections based on current QPS and target QPS.
+   */
+  private class SleepTimeUpdater implements Runnable {
+    @Override
+    public void run() {
+      for (Map.Entry<String, AtomicLong> hostAndRequestCount : hostToRequestCount.entrySet()) {
+        String hostname = hostAndRequestCount.getKey();
+        AtomicLong requestCount = hostAndRequestCount.getValue();
+        Long hostSleepTime = hostToSleepTime.get(hostname);
+        logger.info("For host {}, request count per sec is {}, current sleep time is {} ms", hostname,
+            requestCount.get(), hostSleepTime);
+        if (targetQPS > 0 && requestCount.get() > 0) {
+          long val = requestCount.get() * 100 * hostSleepTime / targetQPS;
+          hostSleepTime = val / 100 + (val > 100 && (val % 100 >= 50 || hostSleepTime == 1) ? 1 : 0);
```

Comment:
I don't really understand your calculation here, can you elaborate?

BTW, I think some simple function like 
```
long rtt = 1000 / request.get() - hostSleepTime;
hostSleepTime -= rrt
```
might just do

Reply:
fixed